### ISSUE TITLE
Default account is renamable

### ISSUE THREAD

**Issue by jrick** [btcsuite/btcwallet#245]
This behavior strikes me as wrong:

```
$ btcctl --simnet --wallet renameaccount default default2    
$ btcctl --simnet --wallet listaccounts
{  
  "default2": 9.99997,  
  "imported": 0  
}  
$ btcctl --simnet --wallet renameaccount default2 ''
-4: account with the same name already exists  
$ btcctl --simnet --wallet renameaccount default2 default
-4: reserved account name  
```

Three problems:

1. The default account is accessible by two names, the empty string and "default".
2. The "default" pseudonym can be renamed to something else, without actually changing the "real" account name, the empty string.
3. Once the default pseudonym has been modified, it cannot be changed back to either "default" (it's a reserved name) or "" (the account still exists).

All three of these problems can be removed by simply always using "" as the identifier of the default account.  Sure, there may be users who don't understand what "" is and why it can't be renamed, but that is a problem that should be solved by a wallet client (cli, gui, etc.)

**Comment by davecgh** [comment]
While I agree that the current logic clearly isn't handling the rename cases properly, I don't want to see the ability rename the default account taken away.  That is not a good user experience.

**Comment by davecgh** [comment]
In response to the specific points:

1) Being accessible by "" and whatever the actual name is the correct behavior in my opinion because the CLI uses "" as an alias for default account.  Therefore specifying either the real name for the default account of "" both make sense.
2) This is likely the root of the issue.  The account should actually be renamed while "" remains as an alias to it.
3) The default account (and only the default account) should be able to renamed back to "" without issue.

**Comment by tuxcanfly** [comment]
I tried to simplify the validation a bit and weed out one issue in `RenameAccount` i.e. we should check validation before duplicate name. 

https://github.com/tuxcanfly/btcwallet/commit/b5cc5630ed893b1c9917d254dab744f0afebdb1c

With this approach we limit reserved names to "", "*" where "" is reserved because it always points to the default account and "*" because it's the keyword for all accounts.

**Comment by jrick** [comment]
The default account should absolutely never be renamed.  Then you'll run into problems with btcjson filling in the default account for you (to the empty string) when it was left unspecified, and nearly every wallet request will begin failing.  If you want to talk about bad UX, this is far worse. 

If you don't like the default account name, make a different account.  Or use a fancy client that gives a pretty name to the default account.  But again, this is a client problem.  I don't want to complicate wallet code for unneeded corner cases like this (it's already caused one bug, wouldn't be surprised if there are more).

**Comment by tuxcanfly** [comment]
Hmm, I agree about complexity but I don't think there should be a problem with client requests. In the above example, even after renaming, queries with empty string continue to work with default account:

    $ btcctl --simnet --wallet renameaccount default default2    
    $ btcctl --simnet --wallet listaccounts
    {  
    "default2": 9.99997,  
    "imported": 0  
    }  
    $ btcctl --simnet --wallet getnewaddress ''    
    Sxxxxxxx
    $ btcctl --simnet --wallet getaddressesbyaccount ''    
    [...]
    $ btcctl --simnet --wallet getaddressesbyaccount default2
    [...] (same as previous)

To put it simply, '' is should be considered a special reserved keyword which always points to the default account. The actual account name could be different (default, default2 whatever)

**Comment by davecgh** [comment]
Another alternative is to simply change btcjson so if doesn't have a default for the account.  It would only entail a change to the struct tag in the struct definitions and updating the associated tests.  Then wallet can examine the account for nil, if it is, it means the default account regardless of its name.  Otherwise, if specified, it has to match the actual name.

However, that said, I prefer the approach outlined by tux, because it will work well when you have to specify things after the account and you want the default account.  '' = default regardless of the real name.  Simple and easy to explain.

Some other cases:

```
$ btcctl ... renameaccount '' default3     ('' = default so it's renamed to default3 regardless of what it was)
$ btcctl ... renameaccount default3 ''     (default account is actually renamed to the empty string)
$ btcctl ... renameaccount non-default ''  (fail, non-default accounts must have a non-empty, unique name)
$ btcctl ... renameaccount '' ''           ('' = default and default can be renamed to empty string, so it works)
```

**Comment by jrick** [comment]
I guess, but I don't see what this solves or how it is an improvement.  It's not a UX concern because these problems go away when you start using a user-friendly GUI or CLI client (sorry, btcctl does *not* count).  Wallet remains simpler (less corner cases to test), clients get very specific behavior that is never surprising, and everyone is happy.

The way I see it is that there are two "kinds" of wallet clients that a user is likely to end up using: ones that only show the default account (and don't support any kind of accounts other than that default account), and big fancy clients that do include accounts.  For the first case, the client is not interested in figuring out which account is the default.  As an example, currently `listaccounts` uses the "fancy" name for the default account.  Because this fancy name can be changed, if it ever is, the client has no idea which account is actually the one it cares about and has been showing the user info for all this time.  For the second case, I doubt these clients would even bother using the default account (it would be unfortunate if a client like the simple one in the first example were to spend money from some important account by accident).  For these clients, where you explicitly track the accounts of every bitcoin coming and going from your wallet, you never want to have the funds in the default account mistaken for something else by another client, so they simply won't bother using the default.

Client behavior should not be broken just because the user came by with another client and renamed an account.  Doing naming tricks with accounts is rather magical and not obvious from a design and API standpoint, and it complicates correct client design.

If we absolutely must have aliases for the default account, at least put that in the RPC server.  This is not a concern of the address manager.

**Comment by tuxcanfly** [comment]
Yup, it's just a naming trick, but it felt like a good balance between keeping compat and allowing multiple accounts. I do agree about the increased complexity and corner cases and as you said, it's particularly imported to not mix-up accounts, so I'm fine with keeping it straightforward as done in #246

**Comment by davecgh** [comment]
So for an update, we hashed this out on IRC.

The commands which accept an account are going to have the btcjson defaults removed (so it doesn't auto populate '' when not specified) and the wallet RPC will return an error like "No account specified.  Use listaccounts to obtain a list of valid accounts." if the default account has been renamed and/or new account(s) have been added.  When the default account hasn't been renamed and there is only one account, the account name will not need to be specified.

**Comment by tuxcanfly** [comment]
OK thanks, that makes sense.
